tools/internal/parser/parser_test.go

```diff
@@ -0,0 +1,777 @@
+package parser
+
+import (
+	"bytes"
+	"cmp"
+	"net/mail"
+	"net/url"
+	"os"
+	"slices"
+	"strings"
+	"testing"
+
+	diff "github.com/google/go-cmp/cmp"
+)
+
+// TestParser runs a battery of synthetic parse and validation tests.
+func TestParser(t *testing.T) {
+	// These test cases have a fair amount of repetition in them,
+	// since both errors and suffix blocks contain repeated nestings
+	// of blocks and Source objects. While it's tempting to try and
+	// reduce duplication through clever code, you are encouraged to
+	// resist the urge.
+	//
+	// Each test case is quite verbose, but being laid out with
+	// minimal indirection makes it easier to inspect and debug when a
+	// failure happens.
+
+	tests := []struct {
+		name               string
+		psl                string
+		downgradeToWarning func(error) bool
+		want               File
+	}{
+		{
+			name: "empty",
+			psl:  "",
+			want: File{},
+		},
+
+		{
+			name: "just_comments",
+			psl: dedent(`
+			  // This is an empty PSL file.
+
+			  // Here is a second comment.
+			`),
+			want: File{
+				Blocks: []Block{
+					Comment{Source: src(1, 1, "// This is an empty PSL file.")},
+					Comment{Source: src(3, 3, "// Here is a second comment.")},
+				},
+			},
+		},
+
+		{
+			name: "just_suffixes",
+			psl: dedent(`
```

Comment:
Indentation weird?

Reply:
Hmm, that's annoying. This is because Go code indents with tabs+spaces (the classic "correct" indentation style for code), but these quoted strings are indented with spaces only. So they look aligned in my editor, with my specific tab width, but not on github.

I replaced the `dedent` helper with a `lines` helper, which uses a list of strings instead of a multiline string. This preserves indentation correctly.